### PR TITLE
fix: limit number of concurrent optimized compactions

### PR DESCRIPTION
**Pull request by gwossum**
Limit number of concurrent optimized compactions so that level compactions do not get starved. Starved level compactions result in a sudden increase in disk usage.

Add [data] max-concurrent-optimized-compactions for configuring maximum number of concurrent optimized compactions. Default value is 1.

Closes: #26315

### CODE REVIEW

**Comment by davidby-influx** [comment]
@gwossum & @devanbenz - I have broken out the compaction planning into a method separate from the asynchronous Go proc with a ticker that drive it. This let me write a test that I think we should massively expand to test and explore the planner.

One thing I think is a bug is that we count generations skipping files in use when we are planning optimization.  This means in my second new test case, the generation count is zero, because the files were acquired in the initial level planning. But I think that the generation count in this case should be one.  What do you think?

Please expand the test cases in  `TestEnginePlanCompactions` and make sure that `Engine.PlanCompactions` is doing the right thing in all the edge cases.

File: tsdb/engine/tsm1/engine.go [hunk 1]
```diff
@@ -2206,13 +2277,79 @@ func (e *Engine) compact(wg *sync.WaitGroup) {
...
+		} else {
+			level4Groups = append(level4Groups, group)
 		}
 	}
+
+	// level5Aggressive indicates if /all/ level 5 compactions should be done at the aggressive points per block, or only those
+	// which have already been done at the aggressive points per block.
+	// If no full compactions are needed, see if an optimize is needed
+	if len(level4Groups) == 0 {
+		plannedLevel5Groups, _, genCount := e.CompactionPlan.PlanOptimize(e.LastModified())
```
Comment (davidby-influx) [inline tsdb/engine/tsm1/engine.go:2319]: I believe we return the wrong generation count here because `DefaultPlanner.PlanOptimize` calls `DefaultPlanner.findGenerations(true)`, skipping files in use.  But, at this point,  files we are looking at are already in a plan. See notes in  `TestEnginePlanCompactions`

File: tsdb/engine/tsm1/compact_test.go [hunk 2]
```diff
@@ -3896,6 +3896,204 @@ func TestDefaultPlanner_Plan_ForceFull(t *testing.T) {
...
+			blockCounts: []int{
+				tsdb.DefaultAggressiveMaxPointsPerBlock,
+				tsdb.DefaultMaxPointsPerBlock,
+				tsdb.DefaultMaxPointsPerBlock,
+				tsdb.DefaultMaxPointsPerBlock,
+			},
+			level5Groups: []tsm1.CompactionGroup{
+				{"01-05.tsm", "02-05.tsm", "03-05.tsm", "04-05.tsm"},
+			},
+			level5Aggressive: false,
```
Comment (davidby-influx) [inline tsdb/engine/tsm1/compact_test.go:4055]: I believe that this should be true, but we get 0 generations back from `PlanOptimize` because we skip files in use.  Please step through the second run of this test and see if you agree.

Comment (devanbenz) [reply tsdb/engine/tsm1/compact_test.go:4055]: I believe this should be `true` since we have a file that is at the aggressive block count within these TSM groups. I think there are two things that would need to adjusted. 
1. https://github.com/influxdata/influxdb/pull/26319/files#diff-e0d6c334f09b26e18b81206ad3b242f2fd1ef09a7b23ea326a13eea3ac916a9aR2306-R2309 We will need to `Release` the acquired TSM group here that is already in the `filesInUse` map so they can be re-acquired by the PlanOptimize planner run. 
2. https://github.com/influxdata/influxdb/pull/26319/files#diff-e0d6c334f09b26e18b81206ad3b242f2fd1ef09a7b23ea326a13eea3ac916a9aR2322 This test will need to account for a group that has a TSM file that is already at `aggressiveCompactionPointsPerBlock`

File: tsdb/engine/tsm1/engine.go [hunk 3]
```diff
@@ -2206,13 +2277,79 @@ func (e *Engine) compact(wg *sync.WaitGroup) {
...
+	// to maintain optimization for the entire group to avoid "going backwards" on the
+	// optimization level. For instance, if an optimized cold shard had back-fill data
+	// added to it, we should maintain the optimization to avoid unoptimizing the bulk of
+	// the shards only to need to reoptimize them later.
+	// In an ideal world, CompactionPlan.Plan and CompactionPlan.PlanOptimize might handle this.
+	level4Groups = make([]CompactionGroup, 0, len(initialLevellevel4Groups))
+	level5Groups = make([]CompactionGroup, 0, len(initialLevellevel4Groups))
+	for _, group := range initialLevellevel4Groups {
+		if isOpt, filename, heur := e.IsGroupOptimized(group); isOpt {
+			e.logger.Info("Promoting full compaction level 4 group to optimized level 5 compaction group because it contains an already optimized TSM file",
```
Comment (devanbenz) [inline tsdb/engine/tsm1/engine.go:2307]: Lower in the code we check to see if the level 5 plan should be set to aggressive. If we already have a TSM file set to the aggressive blocks count we should probably have `level5Aggressive = true`?

File: tsdb/engine/tsm1/engine.go [hunk 4]
```diff
@@ -2206,13 +2277,79 @@ func (e *Engine) compact(wg *sync.WaitGroup) {
...
+				zap.String("optimized_file", filename), zap.String("heuristic", heur), zap.Strings("files", group))
+			level5Groups = append(level5Groups, group)
```
Comment (devanbenz) [inline tsdb/engine/tsm1/engine.go:2309]: We should probably `Release` the group that we moved from `initiallevelLevel4Groups` to `level5Groups` here? So it is no longer in `filesInUse` by the compaction planner and can be picked up by the PlanOptimize planner?

**Review by devanbenz — CHANGES_REQUESTED** [review]
Just two comments regarding the level5Aggressive variable and the logic. As you point out here: https://github.com/influxdata/influxdb/pull/26319/files#r2061085298 it's likely that we have a bug.

**Comment by devanbenz** [comment]
I would really like to find some ways to pull out the actual "scheduling" of compactions too and test some more of this large chunk of code:

https://github.com/influxdata/influxdb/blob/gw/26315/opt_compact_limiter/tsdb/engine/tsm1/engine.go#L2208-L2273

I like how we have modularized planning and made it very easy to test. Would be nice to break some of this out and make it more test-able. Going to start brainstorming on this.

File: tsdb/engine/tsm1/engine.go (hunk 4 above)
Comment (davidby-influx) [reply tsdb/engine/tsm1/engine.go:2309]: That means we should set FileInUse to false when counting generations.

File: tsdb/engine/tsm1/engine.go [hunk 5]
```diff
@@ -2273,63 +2263,138 @@ func (e *Engine) compact(wg *sync.WaitGroup) {
...
-	level2Groups, len2 := e.CompactionPlan.PlanLevel(2)
-	level3Groups, len3 := e.CompactionPlan.PlanLevel(3)
+	l1Groups, len1 := e.CompactionPlan.PlanLevel(1)
+	l2Groups, len2 := e.CompactionPlan.PlanLevel(2)
+	l3Groups, len3 := e.CompactionPlan.PlanLevel(3)
 	initialLevellevel4Groups, _ := e.CompactionPlan.Plan(e.LastModified())
 
+	for _, group := range l1Groups {
+		level1Groups = append(level1Groups, PlannedCompactionGroup{
+			Group: group,
```
Comment (davidby-influx) [inline tsdb/engine/tsm1/engine.go:2333]: No points per block needed here?

File: tsdb/engine/tsm1/engine.go [hunk 6]
```diff
@@ -2273,63 +2263,138 @@ func (e *Engine) compact(wg *sync.WaitGroup) {
...
-			e.logger.Info("Planned optimized level 5 compactions are multi-generational. Only already aggressive groups will use aggressive points per block.")
-		}
-
-		// Put the planned optimize compaction groups after any promoted full compaction groups.
 		for _, group := range plannedLevel5Groups {
-			e.logger.Info("Planning optimized level 5 compaction group", zap.Strings("files", group), zap.Bool("aggressive", level5Aggressive))
-			level5Groups = append(level5Groups, group)
+			// If a level5 optimized compaction group is a single generation. We will need to rewrite
+			// the files at a higher points per block count in order to fully compact them in to a single TSM file.
+			if genCount == 1 {
```
Comment (davidby-influx) [inline tsdb/engine/tsm1/engine.go:2385]: I'm still unclear whether we should set `filesInUse = false` in the call to `findGenerations` whihch produces this.  The `initialLevellevel4Groups` will be marked as in use, so we don't get an accurate generation count of the files we care about.

**Review by davidby-influx — CHANGES_REQUESTED** [review]
Two questions, still need to review test coverage.

File: tsdb/engine/tsm1/engine.go (hunk 6 above)
Comment (devanbenz) [reply tsdb/engine/tsm1/engine.go:2386]: In 
```go
	level4Groups = make([]PlannedCompactionGroup, 0, len(initialLevellevel4Groups))
	level5Groups = make([]PlannedCompactionGroup, 0, len(initialLevellevel4Groups))
	for _, group := range initialLevellevel4Groups {
		if isOpt, filename, heur := e.IsGroupOptimized(group); isOpt {
			e.logger.Info("Promoting full compaction level 4 group to optimized level 5 compaction group because it contains an already optimized TSM file",
				zap.String("optimized_file", filename), zap.String("heuristic", heur), zap.Strings("files", group))

			// Should set this compaction group to aggressive. IsGroupOptimized will check the
			// block count and return true if there is a file at aggressivePointsPerBlock.
			// We will need to run aggressive compaction on this group if that's the case.
			level5Groups = append(level5Groups, PlannedCompactionGroup{
				Group:          group,
				PointsPerBlock: e.CompactionPlan.GetAggressiveCompactionPointsPerBlock(),
			})
		} else {
			level4Groups = append(level4Groups, PlannedCompactionGroup{
				Group:          group,
				PointsPerBlock: tsdb.DefaultMaxPointsPerBlock,
			})
		}
	}
```

We move the level4Groups over to level5Groups if any of the TSM files are at the aggressive points per block count. I don't really think we should be 're-running' that group through the PlanOptimize compaction planner :thinking: If we were to basically clear out the files in use (or remove this group that we are now appending to level5) we would effectively rerun it through the planner which would just be extra work. I'm open to doing that way but I don't see the benefit since we have already established this group should be ran through the optimized compaction with a larger points per block.

**Review by davidby-influx — COMMENTED** [review]
Comments, some outdated.

**Review by davidby-influx — DISMISSED** [review]
LGTM.  Please wait for @gwossum's approval, as well before merging.

File: tsdb/engine/tsm1/engine.go [hunk 7]
```diff
@@ -2096,11 +2118,39 @@ func (e *Engine) ShouldCompactCache(t time.Time) bool {
...
+			return true, f, heur
+		}
+	}
+	return false, "", ""
+}
+
+const waitForOptimization = time.Hour
+const tickPeriod = time.Second
+
+var ticksBeforeOptimize = int(waitForOptimization.Seconds())
```
Comment (gwossum) [inline tsdb/engine/tsm1/engine.go:2145]: `ticksBeforeOptimize` is only correct if `tickPeriod` is `time.Second`.

I think this will generalize it if `tickPeriod` is not `time.Second`:
```go
var ticksBeforeOptimize = int(waitForOptimization.Seconds() * time.Second / tickPeriod)
```

File: tsdb/engine/tsm1/engine.go [hunk 8]
```diff
@@ -2124,31 +2174,16 @@ func (e *Engine) compact(wg *sync.WaitGroup) {
...
-				if len(level4Groups) > 0 {
-					for _, group := range level4Groups {
-						e.logger.Info("TSM scheduled for optimized compaction", zap.Strings("files", group))
-					}
+			cycleCount += 1
+			skipOptimize := func() (bool, string) {
+				if cycleCount <= ticksBeforeOptimize {
+					return true, waitMessage
+				} else {
+					return false, ""
```
Comment (gwossum) [inline tsdb/engine/tsm1/engine.go:2182]: It might be better to do a `startTime := ticks.Now()` before the loop and then do `if time.Since(startTime) > waitForOptimize` here because this loop may not take exactly one `tickPeriod` to run. Also, it seems easier since you don't have to do dimensional analysis to determine if the `ticksBeforeOptimize` calculation is correct.

Comment (devanbenz) [reply tsdb/engine/tsm1/engine.go:2182]: This loop runs indefinitely for the life of the program afaik. Wouldn't start time just always be program start? So after the first time it is greater than waitForOptimize it will always be greater?

**Review by devanbenz — APPROVED** [review]
LGTM

File: tsdb/engine/tsm1/engine.go (hunk 8 above)
Comment (davidby-influx) [reply tsdb/engine/tsm1/engine.go]: Yes.  This just avoids optimizing compactions for the first hour after start
